Add OptionalTensorRef; `nll_loss_forward`: port to structured kernel

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61443**

For more information, see #55070.

This PR also adds a new type, `OptionalTensorRef` as a replacement for `c10::optional<Tensor>&` in order to avoid the reference count manipulations that are inevitable with the latter. I have confirmed using Godbolt/Compiler Explorer that this class does indeed avoid manipulating the reference count of the `intrusive_ptr` inside the `Tensor` it refers to:

1. [P429709479](https://www.internalfb.com/phabricator/paste/view/P429709479) - Given a `const Tensor&` in scope, an `OptionalTensorRef` can be constructed without bumping refcount.
2. [P429709883](https://www.internalfb.com/phabricator/paste/view/P429709883) - Given an `OptionalTensorRef`, a `const Tensor&` can be produced without bumping refcount.
3. [P429710335](https://www.internalfb.com/phabricator/paste/view/P429710335) - When `OptionalTensorRef` is destructed, the refcount should not be decremented.
4. [P429769525](https://www.internalfb.com/phabricator/paste/view/P429769525) - `OptionalTensorRef` can be assigned without refcount manipulation.
5. [P429769882](https://www.internalfb.com/phabricator/paste/view/P429769882) - `OptionalTensorRef` can be move assigned without refcount manipulation.

Differential Revision: [D29780666](https://our.internmc.facebook.com/intern/diff/D29780666)

Commit:
[WIP] `nll_loss_forward`: port to structured kernel

For more information, see #55070.

[ghstack-poisoned]

Commit:
Update on "[WIP] `nll_loss_forward`: port to structured kernel"


For more information, see #55070.

[ghstack-poisoned]

Commit:
Update on "[WIP] `nll_loss_forward`: port to structured kernel"


For more information, see #55070.

[ghstack-poisoned]

Commit:
Update on "[WIP] `nll_loss_forward`: port to structured kernel"


For more information, see #55070.

[ghstack-poisoned]

Commit:
Update on "[WIP] `nll_loss_forward`: port to structured kernel"


For more information, see #55070.

[ghstack-poisoned]

Commit:
Update on "[WIP] `nll_loss_forward`: port to structured kernel"


For more information, see #55070.

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/Tensor.h:
```diff
@@ -1,3 +1,33 @@
+  OptionalTensorRef(const Tensor& src)
+      : ref_(c10::intrusive_ptr<TensorImpl>(
+            src.unsafeGetTensorImpl(),
+            c10::raw::DontIncreaseRefcount{})) {}
```
For extra safety, assert src is NOT undefined here.

ezyang replied:
I don't see the assert here?

SplitInfinity replied:
Oops, got deleted in my last revision for some reason. I'll add it back.

ezyang commented on tools/codegen/api/translate.py:
```diff
@@ -94,6 +94,12 @@ def translate(
             ctx[NamedCType(t.elem.elem.name, ConstRefCType(BaseCType(tensorT)))] = \
                 f'({b.expr}.has_value() ? *{b.expr} : at::Tensor())'
```
Oh interesting. Actually, the `isinstance(t, NamedCType)` check below is not necessary, because the code above is dead. You can tell this because Binding's type is NamedCType.

ezyang commented on tools/codegen/api/translate.py:
```diff
@@ -94,6 +94,12 @@ def translate(
 
+        if isinstance(t, NamedCType) and isinstance(t.type, ConstRefCType) and \
+                isinstance(t.type.elem, OptionalCType) and isinstance(t.type.elem.elem, BaseCType) \
+                and str(t.type.elem.elem.type) == 'at::Tensor':
```
I know you're copying the code above but you can actually do this much more straightforwardly: `t.type == ConstRefCType(OptionalCType(BaseCType(tensorT)))`

ezyang commented on aten/src/ATen/native/LossNLL.cpp:
```diff
@@ -2,10 +2,48 @@
+      " but got weight tensor of shape: ",
+      weight.sizes());
+
+  set_output(0, {0}, self.options());
```
This looks pretty dodgy, surely the zeroth output isn't always a zero size tensor.

SplitInfinity replied:
Given that this kernel is computing loss, I think the output is a scalar? This is the old implementation of the non-out variant:

```
  auto output = at::empty({0}, self.options());
  auto total_weight = at::empty({0}, self.options());
  at::native::nll_loss_forward_out_cpu(
      self, target, weight, reduction, ignore_index, output, total_weight);
  return std::make_tuple(output, total_weight);
```

The `{}` size for the second output comes from the unconditional `total_weight.resize_({})` in the old implementation `nll_loss_forward_out_cpu_template`.

ezyang replied:
If the output was a scalar, the output size would be specified as `{}`, not `{0}`. My guess is `nll_loss_forward_out_cpu` is internally doing a resize.

SplitInfinity replied:
You're right, I found the code that resizes `output` in `nll_loss_out_frame` (which is called by `nll_loss_forward_out_cpu_template`). All the other kernels I've ported did all their resizing in `{kernel_name}_cpu_template`, not `{kernel_name}_frame` so I didn't think to look there. I'll fix in a bit after doing some PSC

Commit:
Update on "`nll_loss_forward`: port to structured kernel"


For more information, see #55070.

This PR also adds a new type, `OptionalTensorRef` as a replacement for `c10::optional<Tensor>&` in order to avoid the reference count manipulations that are inevitable with the latter. I have confirmed using Godbolt/Compiler Explorer that this class does indeed avoid manipulating the reference count of the `intrusive_ptr` inside the `Tensor` it refers to:

1. [P428856781](https://www.internalfb.com/phabricator/paste/view/P428856781) - Given a `const Tensor&` in scope, an `OptionalTensorRef` can be constructed without bumping refcount.
2. [P428857411](https://www.internalfb.com/phabricator/paste/view/P428857411) - Given an `OptionalTensorRef`, a `const Tensor&` can be produced without bumping refcount.
3. [P428860300](https://www.internalfb.com/phabricator/paste/view/P428860300) - When `OptionalTensorRef` is destructed, the refcount should not be decremented.

[ghstack-poisoned]

SplitInfinity commented on c10/util/intrusive_ptr.h:
```diff
@@ -285,9 +285,11 @@ class intrusive_ptr final {
   // This constructor will not increase the ref counter for you.
   // We use the tagged dispatch mechanism to explicitly mark this constructor
   // to not increase the refcount
+ public:
```
We should talk about this. I think we need this constructor for `OptionalTensorRef` to work as intended but it must be private for a good reason.

ezyang replied:
I'm +1 for this constructor, it is tagged (so you have to explicitly use it) and I am going to claim there is no chesterton fence here.

ezyang replied:
This seems like it should work. Are the pastes out of date?

SplitInfinity replied:
No, they aren't out of date. I didn't realize the refcount increments/decrements were inlined by `-O2` and show up in the assembly as `lock xaddq %rdx, 8(%rax)`, but after playing around with my code in the compiler explorer some more, I realized that the refcount increment that is visible in the assembly is from copying the `OptionalTensorRef` instance when calling `externalFunctionOptionalTensorRef`:
```
extern void externalFunctionOptionalTensorRef(OptionalTensorRef ref);

/**
Requirement #1
Given a const Tensor& in scope, produce an OptionalTensor (or reference to optional tensor) without bumping refcount.
*/
void createOptionalTensorRefFromTensor(const Tensor& src) {
    OptionalTensorRef refToSrc(src);
    externalFunctionOptionalTensorRef(refToSrc);   // refToSrc is *copied* to the parameter named ref
}
```
There is no corresponding refcount decrement because `~OptionalTensorRef` calls `release_()`. Adding a copy constructor for `OptionalTensorRef` that constructs `this->ref_` using this constructor of `c10::intrusive_ptr` that does not increase reference count solves the problem. I'll update the PR tomorrow.

SplitInfinity replied:
I updated the pastes with the new assembly.

Commit:
Update on "`nll_loss_forward`: port to structured kernel"


For more information, see #55070.

This PR also adds a new type, `OptionalTensorRef` as a replacement for `c10::optional<Tensor>&` in order to avoid the reference count manipulations that are inevitable with the latter. I have confirmed using Godbolt/Compiler Explorer that this class does indeed avoid manipulating the reference count of the `intrusive_ptr` inside the `Tensor` it refers to:

1. [P429071558](https://www.internalfb.com/phabricator/paste/view/P429071558) - Given a `const Tensor&` in scope, an `OptionalTensorRef` can be constructed without bumping refcount.
2. [P429071968](https://www.internalfb.com/phabricator/paste/view/P429071968) - Given an `OptionalTensorRef`, a `const Tensor&` can be produced without bumping refcount.
3. [P429071423](https://www.internalfb.com/phabricator/paste/view/P429071423) - When `OptionalTensorRef` is destructed, the refcount should not be decremented.

[ghstack-poisoned]

ezyang commented:
No, it doesn't seem like refcount manipulation has been avoided in your pastes. For example there is a locked instruction here https://www.internalfb.com/phabricator/paste/view/P429071558?lines=87 and here https://www.internalfb.com/phabricator/paste/view/P429071423?lines=85

ezyang commented on aten/src/ATen/core/Tensor.h:
```diff
@@ -1,3 +1,36 @@
+      : ref_(c10::intrusive_ptr<TensorImpl>(
+            src.unsafeGetTensorImpl(),
+            c10::raw::DontIncreaseRefcount{})) {
+    TORCH_INTERNAL_ASSERT(src.defined());
```
guessing better for this to be DEBUG_ONLY

ezyang commented on aten/src/ATen/native/LossNLL.cpp:
```diff
@@ -2,10 +2,56 @@
+  const auto batch_size = self.size(0);
+
+  if (reduction == Reduction::None && n_dims == 2) {
+    set_output(0, {batch_size}, self.options());
```
I was expecting to see a deletion of the `resize_` that handles this case, but I don't see it. Where is it?

SplitInfinity replied:
Lol I forgot to delete them 🙃 🔫

Commit:
Update on "`nll_loss_forward`: port to structured kernel"


For more information, see #55070.

This PR also adds a new type, `OptionalTensorRef` as a replacement for `c10::optional<Tensor>&` in order to avoid the reference count manipulations that are inevitable with the latter. I have confirmed using Godbolt/Compiler Explorer that this class does indeed avoid manipulating the reference count of the `intrusive_ptr` inside the `Tensor` it refers to:

1. [P429071558](https://www.internalfb.com/phabricator/paste/view/P429071558) - Given a `const Tensor&` in scope, an `OptionalTensorRef` can be constructed without bumping refcount.
2. [P429071968](https://www.internalfb.com/phabricator/paste/view/P429071968) - Given an `OptionalTensorRef`, a `const Tensor&` can be produced without bumping refcount.
3. [P429071423](https://www.internalfb.com/phabricator/paste/view/P429071423) - When `OptionalTensorRef` is destructed, the refcount should not be decremented.

[ghstack-poisoned]

Commit:
Update on "`nll_loss_forward`: port to structured kernel"


For more information, see #55070.

This PR also adds a new type, `OptionalTensorRef` as a replacement for `c10::optional<Tensor>&` in order to avoid the reference count manipulations that are inevitable with the latter. I have confirmed using Godbolt/Compiler Explorer that this class does indeed avoid manipulating the reference count of the `intrusive_ptr` inside the `Tensor` it refers to:

1. [P429709479](https://www.internalfb.com/phabricator/paste/view/P429709479) - Given a `const Tensor&` in scope, an `OptionalTensorRef` can be constructed without bumping refcount.
2. [P429709883](https://www.internalfb.com/phabricator/paste/view/P429709883) - Given an `OptionalTensorRef`, a `const Tensor&` can be produced without bumping refcount.
3. [P429710335](https://www.internalfb.com/phabricator/paste/view/P429710335) - When `OptionalTensorRef` is destructed, the refcount should not be decremented.

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/Tensor.h:
```diff
@@ -1,3 +1,41 @@
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(src.defined());
+  }
+
+  OptionalTensorRef(const OptionalTensorRef& rhs)
```
This isn't enough. You need to define copy assignment as well. And while we're at it, it's probably safest to explicitly define what the move constructor and move assignment do (though I think the defaults do what you want)

ezyang replied:
(I think by manually defining the move constructor/assignment you can also apply a tiny optimization which is that you don't have to clear the source OptionalTensorRef on a move; it's non owning!)

SplitInfinity replied:
I added a move constructor/assignment operator but the generated assembly for a move assignment statement (see PR description) is exactly the same as what get I don't declare them at all. Interestingly enough, if I *do* declare them with `= default`, the assembly does change (and there are refcount manipulations involved).

It could be that I'm not feeding the right code into the compiler but I'm inclined to get rid of them since the implicitly generated ones seem to be doing the same thing as my implementations.

SplitInfinity replied:
Or maybe I implemented them incorrectly? I'm not clearing the source as far as I can tell...

ezyang replied:
Hmm, so what it seems is happening is that when you explicitly declare copy constructor, this deletes the move constructor. So actually whenever a move would have been possible, the C++ compiler does a copy instead. Which normally would be a pessimization, but in this case it's fine! 

Prompted by this, I checked c++17 `string_view` to see what they'd done, and they had omitted them move assignment/constructor. So in fact you got it the right the first try and you should ignore me...

ezyang replied:
(To be clear, you still need to define copy assignment; that's rule of three https://en.cppreference.com/w/cpp/language/rule_of_three )

SplitInfinity replied:
Okay, I deleted the move constructor and assignment operator.

ezyang commented on aten/src/ATen/core/Tensor.h:
```diff
@@ -1,3 +1,41 @@
+
+  const Tensor& getTensorRef() const {
+    return ref_;
+  }
```
This needs the ref qualifier as described in the RFC at https://github.com/pytorch/rfcs/pull/16/files#diff-04c8518032fa82b79c85ae7e3c210d7c9ef27bc103484cbc85728d79b42f2d6aR47 since `const Tensor&` will not induce const reference lifetime extension see https://github.com/pytorch/rfcs/pull/16/files#diff-04c8518032fa82b79c85ae7e3c210d7c9ef27bc103484cbc85728d79b42f2d6aR83-R88

SplitInfinity replied:
Sigh, details are hard.

Commit:
Update on "`nll_loss_forward`: port to structured kernel"


For more information, see #55070.

This PR also adds a new type, `OptionalTensorRef` as a replacement for `c10::optional<Tensor>&` in order to avoid the reference count manipulations that are inevitable with the latter. I have confirmed using Godbolt/Compiler Explorer that this class does indeed avoid manipulating the reference count of the `intrusive_ptr` inside the `Tensor` it refers to:

1. [P429709479](https://www.internalfb.com/phabricator/paste/view/P429709479) - Given a `const Tensor&` in scope, an `OptionalTensorRef` can be constructed without bumping refcount.
2. [P429709883](https://www.internalfb.com/phabricator/paste/view/P429709883) - Given an `OptionalTensorRef`, a `const Tensor&` can be produced without bumping refcount.
3. [P429710335](https://www.internalfb.com/phabricator/paste/view/P429710335) - When `OptionalTensorRef` is destructed, the refcount should not be decremented.
4. [P429769525](https://www.internalfb.com/phabricator/paste/view/P429769525) - `OptionalTensorRef` can be assigned without refcount manipulation.
5. [P429769882](https://www.internalfb.com/phabricator/paste/view/P429769882) - `OptionalTensorRef` can be move assigned without refcount manipulation.

[ghstack-poisoned]

Commit:
Update on "`nll_loss_forward`: port to structured kernel"


For more information, see #55070.

This PR also adds a new type, `OptionalTensorRef` as a replacement for `c10::optional<Tensor>&` in order to avoid the reference count manipulations that are inevitable with the latter. I have confirmed using Godbolt/Compiler Explorer that this class does indeed avoid manipulating the reference count of the `intrusive_ptr` inside the `Tensor` it refers to:

1. [P429709479](https://www.internalfb.com/phabricator/paste/view/P429709479) - Given a `const Tensor&` in scope, an `OptionalTensorRef` can be constructed without bumping refcount.
2. [P429709883](https://www.internalfb.com/phabricator/paste/view/P429709883) - Given an `OptionalTensorRef`, a `const Tensor&` can be produced without bumping refcount.
3. [P429710335](https://www.internalfb.com/phabricator/paste/view/P429710335) - When `OptionalTensorRef` is destructed, the refcount should not be decremented.
4. [P429769525](https://www.internalfb.com/phabricator/paste/view/P429769525) - `OptionalTensorRef` can be assigned without refcount manipulation.
5. [P429769882](https://www.internalfb.com/phabricator/paste/view/P429769882) - `OptionalTensorRef` can be move assigned without refcount manipulation.

[ghstack-poisoned]

Commit:
Update on "`nll_loss_forward`: port to structured kernel"


For more information, see #55070.

This PR also adds a new type, `OptionalTensorRef` as a replacement for `c10::optional<Tensor>&` in order to avoid the reference count manipulations that are inevitable with the latter. I have confirmed using Godbolt/Compiler Explorer that this class does indeed avoid manipulating the reference count of the `intrusive_ptr` inside the `Tensor` it refers to:

1. [P429709479](https://www.internalfb.com/phabricator/paste/view/P429709479) - Given a `const Tensor&` in scope, an `OptionalTensorRef` can be constructed without bumping refcount.
2. [P429709883](https://www.internalfb.com/phabricator/paste/view/P429709883) - Given an `OptionalTensorRef`, a `const Tensor&` can be produced without bumping refcount.
3. [P429710335](https://www.internalfb.com/phabricator/paste/view/P429710335) - When `OptionalTensorRef` is destructed, the refcount should not be decremented.
4. [P429769525](https://www.internalfb.com/phabricator/paste/view/P429769525) - `OptionalTensorRef` can be assigned without refcount manipulation.
5. [P429769882](https://www.internalfb.com/phabricator/paste/view/P429769882) - `OptionalTensorRef` can be move assigned without refcount manipulation.

[ghstack-poisoned]

SplitInfinity commented:
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29780666).
